services/replicator/replicator.go

```diff
@@ -932,6 +932,215 @@ func (r *Replicator) CreateRemoteExtent(ctx thrift.Context, createRequest *share
+
+	var err error
+
+	if request == nil || !request.IsSetConsumerGroupUUID() || !request.IsSetExtentUUID() {
+		r.m3Client.IncCounter(metrics.ReplicatorSetAckOffsetInRemoteScope, metrics.ReplicatorFailures)
+		err = &shared.BadRequestError{Message: fmt.Sprintf(`Set ack offset request invalid. IsSetConsumerGroupUUID: [%v], IsSetExtentUUID: [%v]`,
+			request.IsSetConsumerGroupUUID(), request.IsSetExtentUUID())}
+		r.logger.WithField(common.TagErr, err).Error(`Set ack offset request verification failed`)
+		return err
+	}
+
+	lcllg := r.logger.WithFields(bark.Fields{
+		common.TagCnsm: common.FmtDst(request.GetConsumerGroupUUID()),
+		common.TagExt:  common.FmtExt(request.GetExtentUUID()),
+	})
+
+	readCgRequest := shared.ReadConsumerGroupRequest{
+		ConsumerGroupUUID: common.StringPtr(request.GetConsumerGroupUUID()),
+	}
+	cgDesc, err := r.metaClient.ReadConsumerGroup(nil, &readCgRequest)
+	if err != nil {
+		lcllg.WithField(common.TagErr, err).Error(`Error reading cg from metadata`)
+		r.m3Client.IncCounter(metrics.ReplicatorSetAckOffsetInRemoteScope, metrics.ReplicatorFailures)
+		return err
+	}
+
+	if !cgDesc.GetIsMultiZone() {
+		err = &shared.BadRequestError{Message: fmt.Sprintf(`Consumer group [%v] is not multi zone destination`, request.GetConsumerGroupUUID())}
+		lcllg.WithField(common.TagErr, err).Error(`Consumer group is not multi zone destination`)
+		r.m3Client.IncCounter(metrics.ReplicatorSetAckOffsetInRemoteScope, metrics.ReplicatorFailures)
+		return err
+	}
+
+	if !cgDesc.IsSetZoneConfigs() || len(cgDesc.GetZoneConfigs()) == 0 {
+		err = &shared.BadRequestError{Message: fmt.Sprintf(`Zone config for consumer group [%v] is not set`, request.GetConsumerGroupUUID())}
+		lcllg.WithField(common.TagErr, err).Error(`Zone config for consumer group is not set`)
+		r.m3Client.IncCounter(metrics.ReplicatorSetAckOffsetInRemoteScope, metrics.ReplicatorFailures)
+		return err
+	}
+
+	for _, zoneConfig := range cgDesc.GetZoneConfigs() {
+		// skip local zone
+		if strings.EqualFold(zoneConfig.GetZone(), r.localZone) {
+			continue
+		}
+
+		// only forward the call to remote zone if consumer group is visible in that zone
+		if !zoneConfig.GetVisible() {
+			continue
+		}
+
+		// call remote replicators in a goroutine. Errors can be ignored since reconciliation will fix the inconsistency eventually
+		go r.setAckOffsetRemoteCall(zoneConfig.GetZone(), lcllg, request)
+	}
+
+	return nil
+}
+
+func (r *Replicator) setAckOffsetRemoteCall(zone string, logger bark.Logger, request *shared.SetAckOffsetRequest) error {
+	// acquire remote zone replicator thrift client
+	client, err := r.replicatorclientFactory.GetReplicatorClient(zone)
+	if err != nil {
+		r.m3Client.IncCounter(metrics.ReplicatorSetAckOffsetInRemoteScope, metrics.ReplicatorFailures)
+		logger.WithField(common.TagErr, err).Error(`Get remote replicator client failed`)
+		return err
+	}
+
+	// send to remote zone replicator
+	ctx, cancel := thrift.NewContext(remoteReplicatorCallTimeOut)
```

Comment:
consider making this timeout value same as the output host ack offset flush interval (which is 5 seconds, I think)
there is no point in blocking on this call, if the data becomes stale every 5 seconds.